eth/protocols/snap/sync.go

```diff
@@ -0,0 +1,718 @@
+		peer.Log().Warn("Unexpected storage range packet")
+		s.lock.Unlock()
+		return nil
+	}
+	delete(s.storageReqs, peer.id)
+	s.lock.Unlock()
+
+	// If the response is unavailable snapshot, forward to the requester
+	if len(hashes) == 0 && len(slots) == 0 && len(proof) == 0 {
+		select {
+		//case <-req.cancel:
+		case s.storageResps <- nil:
+		}
+		return nil
+	}
+	// Reconstruct a partial trie from the response and verify it
+	keys := make([][]byte, len(hashes))
+	for i, key := range hashes {
+		keys[i] = common.CopyBytes(key[:])
+	}
+	nodes := make(light.NodeList, len(proof))
+	for i, node := range proof {
+		nodes[i] = node
+	}
+	var (
+		db  ethdb.KeyValueStore
+		err error
+	)
+	if len(nodes) == 0 {
+		// No proof has been attached, the response must cover the entire key
+		// space and hash to the origin root.
+		db, _, err = trie.VerifyRangeProof(req.root, req.origin[:], keys, slots, nil, nil)
+		if err != nil {
+			return err
+		}
+	} else {
+		// A proof was attached, the response is only partial, check that the
+		// returned data is indeed part of the storage trie
+		proofdb := nodes.NodeSet()
+
+		db, _, err = trie.VerifyRangeProof(req.root, req.origin[:], keys, slots, proofdb, proofdb)
+		if err != nil {
+			return err
+		}
+	}
+	// Partial trie reconstructed, send it to the scheduler for storage filling
+	bounds := make(map[common.Hash]struct{})
+
+	hasher := sha3.NewLegacyKeccak256()
+	for _, node := range proof {
+		hasher.Reset()
+		hasher.Write(node)
+		bounds[common.BytesToHash(hasher.Sum(nil))] = struct{}{}
+	}
+	last := req.origin
+	if len(hashes) > 0 {
+		last = hashes[len(hashes)-1]
+	}
+	response := &storageResponse{
+		hashes: hashes,
+		slots:  slots,
+		nodes:  db,
+		bounds: bounds,
+		last:   last,
+	}
+	select {
+	//case <-req.cancel:
+	case s.storageResps <- response:
+	}
+	return nil
+}
+
+// OnByteCodes is a callback method to invoke when a batch of contract
+// bytes codes are received from a remote peer.
+func (s *Syncer) OnByteCodes(peer *Peer, id uint64, bytecodes [][]byte) error {
+	peer.Log().Trace("Delivering set of bytecodes", "bytecodes", len(bytecodes))
+
+	// If the request is stale, discard it
+	s.lock.Lock()
+	req, ok := s.bytecodeReqs[peer.id]
+	if !ok || req.id != id {
+		peer.Log().Warn("Unexpected bytecode packet")
+		s.lock.Unlock()
+		return nil
+	}
+	delete(s.storageReqs, peer.id)
+	s.lock.Unlock()
+
+	// Cross reference the requested bytecodes with the response to find gaps
+	// that the serving node is missing
+	hasher := sha3.NewLegacyKeccak256()
+
+	codes := make([][]byte, 0, len(req.hashes))
+	for i, j := 0, 0; i < len(bytecodes); i++ {
+		// Find the next hash that we've been served, add gaps in between
+		hasher.Reset()
+		hasher.Write(bytecodes[i])
+		hash := hasher.Sum(nil)
+
+		for j < len(req.hashes) && !bytes.Equal(hash, req.hashes[j][:]) {
+			codes = append(codes, nil)
+			j++
+		}
+		if j < len(req.hashes) {
+			codes = append(codes, bytecodes[i])
+			j++
+			continue
+		}
```

Comment:
This looks very fishy.... 
So if the `0th` code mismatches. Then codes will be `codes=[nil], j=1`. Then you check `hash` against `req.hashes[1]`.  So for every requested hash, you're searching forward.. 

So it looks like you _almost_ allow codes to be returned in arbitrary order. But if that's the case, I think you're missing half the search-space?

But if you expect the hashes to be in the same order as the request, then it looks very odd that you don't update the `hash` when you move forward among the returned codes.